Added an 'index.html' file to represent the index of all CLI artifacts.

klueska opened the pull request:
This PR adds files to the `scripts` folder for publishing a list of all CLI artifacts to S3. The artifacts themselves are scraped from `downloads.dcos.io/cli/releases` and published as an `artifacts.json` file that is then read and rendered by a small html/css/javascript app.

A live version of the index can be seen here:
https://downloads.dcos.io/cli/index.html

Two different script are used to publish the necessary artifacts for this index.

The first is `publish_index.py` which publishes all of the necessary files required to render the index. This script should not need to be ran very often, and is only necessary if changes are made to any of the files used to implement the rendering of the index.

The second is `publish_artifacts.py` which publishes a JSON file listing all of the artifacts that should be displayed in the index. This script should be ran *every time* we publish a new artifact to S3 which we want to publish as a new release. This applies to both CLI releases themselves, as well as plugins.

In a subsequent commit we should add the calling of `publish_artifacts.py` to our CI as a step in publishing any of our release artifacts.

klueska committed:
Added an 'index.html' file to represent the index of all CLI artifacts.

For now, this file is unusable on it's own. In a subsequent commit, we
will add all of the assets necessary for this file to render properly,
as well as a set of scripts to upload this file (and all of its assests)
to S3 so that end users can see it.

klueska committed:
Added the 'jQuery' asset for use by 'index.html'.

The new index of artifacts uses jQuery and Javascript to manipulate the
DOM when deciding which artifacts to display. This commit adds a
downloaded version of jQuery so that we are not restricted to always
pulling it from its remote source when loading the index.

klueska committed:
Added the 'browser.css' asset for use by 'index.html'.

This style sheet specifies the custom layout of the DOM elements in the
artifacts index. For now, it is very simplisitic, but could be expanded
to have more style consistent with our docs pages in the future.

klueska committed:
Added the 'artifacts.js' asset for use by 'index.html'.

This javascript file abstracts a set of functions used to query and
manipulate the various artifacts we are listing in the index. It is
primarliy used by the (soon to come) 'browser.js' file to build out the
file browser in the artifacts index.

klueska committed:
Added the 'browser.js' asset for use by 'index.html'.

This file abstracts all of the logic necessary to build the file browser
displayed in the artifacts index. It relies on 'artifacts.js' to list
and query the artifacts and builds out the UI to interact list them.

klueska committed:
Added a 'publish_index.py' script to publish the artifacts index to S3.

This script should nto be needed very often. It is only necessary if
changes to the 'index.html' or any of its assets (other than
'artifacts.json') change and need to be reuploaded. A subsequent script
will include the logic to generate 'artifacts.json' and upload it so
that the index dynamically discovers what artifacts it should publish
and display them.

klueska committed:
Added 'publish_artifacts.py' to publish the list of CLI artifacts to S3.

This script is in JSON format and should be called anytime we publish a
new artifact to S3 that we want to include in the artifacts index.

klueska committed:
Added a .gitignore for the 'artifacts.json' asset from 'index.html'.

This JSON file is autogenerated and uploaded separately from the core
files used to display the artifacts index.

armandgrillet commented on scripts/index/index.html:
```diff
@@ -0,0 +1,44 @@
+</head>
+
+<body>
+<h1 id="header"></h1>
```
We should uses classes instead of IDs for `header`, `names`, `versions`, and `listings` as they are used multiple times in this page.

klueska replied:
They really need to be uniquely identifiable though. I could give them unique names if that's clearer. I just always reference them as e.g. `#binaries #header` in CSS or JS code though. Otherwise it would look like `#binaries #binaries-header`.

klueska replied:
What if I were to use "name" like in this answer:
https://stackoverflow.com/questions/9680037/how-do-i-select-an-element-with-its-name-attribute-in-jquery

Is that better?

I don't want to use "class" because I normally think of classes being used to apply style, which is not the case here.

armandgrillet replied:
Yes, `name` would be a better fit.

armandgrillet commented on scripts/index/html/artifacts.js:
```diff
@@ -0,0 +1,354 @@
+      // List unique major versions across all binaries.
+      this.majorVersions = function() {
+        var versions = [];
+        $.each(this.binaries, function(_, binary) {
```
All occurrences of `$.each(a, function(b, c) {` should be replaceable by `a.forEach(function(c, b){`

klueska replied:
Is there a preference? I've often seen $.each() used all over the place in favor of forEach, but I'm not sure why. Can you explain and / or provide a link?

armandgrillet replied:
It is a preference, for this index jQuery could be dropped as you can see in http://youmightnotneedjquery.com/ (nearly all of the jQuery helper functions now exist in vanilla JS). As it would require refactoring and the JS code of this PR is clean, I will resolve this conversation and I don't think that changing any `$` calls is necessary.

armandgrillet commented on scripts/index/publish_artifacts.py:
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+import os
+import boto3
```
How are we getting `boto3`?

klueska replied:
I'm not sure, but all of Bilal's other python scripts use this too without any sort of obvious virtualenv pulling it in. I assume the CI machines have it builtin.

armandgrillet commented on scripts/index/publish_artifacts.py:
```diff
@@ -0,0 +1,86 @@
+
+import os
+import boto3
+import json
```
The import should happen before `import boto3` and also `import os`. Adding a blank line between the stdlib/3rdparty imports would be great.

armandgrillet reviewed (approved):
LGTM. Using jQuery could have been avoided but it's making the code simpler, having multiple elements with the same ID in `index.html` is not best practice but it doesn't break the code. 
I would ship this PR if it is urgent, otherwise please check my comments.

klueska committed:
Updated JSON source for artifacts index to avoid caching.

We now add a random query string to the end of the JSON source used to
generate the artifacts index to avoiding browsers caching the JSON
source. Without this, it was likely that the index would not be updated
properly for many users when refreshing the page.

klueska commented:
@bamarni do you think this script belongs in this repo or elsewhere? It needs to be run anytime there is a new artifact published for binaries and plugins (including the EE CLI).

bamarni commented:
I'd rather see it in dcos/dcos-cli as it's a more generic cli repo.

klueska commented:
Closed in favor of https://github.com/dcos/dcos-cli/pull/1376
